feat: Replace Black with Ruff for code linting and formatting

EverythingElseWasAlreadyTaken opened the pull request:
Replace Black with Ruff for code linting and formatting.
 Run Ruff formatter and apply linting fixes. 
 Remove some unneeded code, that I found while fixing linter issues. 
 Replace the Black workflow with pre-commit workflow. 
Update README.

KelvinChung2000 commented on ruff.toml:
```diff
@@ -0,0 +1,77 @@
+# Like Black, indent with spaces, rather than tabs.
+indent-style = "space"
+
+# Like Black, respect magic trailing commas.
```
Just minor comment mismatch.

EverythingElseWasAlreadyTaken replied:
> Just minor comment mismatch.

The naming of this parameter is a bit ambiguous, but skip-magic-trailing-comma=false, means that it's respected.
https://docs.astral.sh/ruff/settings/#format_skip-magic-trailing-comma

KelvinChung2000 commented on ruff.toml:
```diff
@@ -0,0 +1,77 @@
+# reStructuredText code/literal blocks and doctests are all supported.
+#
+# This is currently disabled by default, but it is planned for this
+# to be opt-out in the future.
```
same here

EverythingElseWasAlreadyTaken replied:
Oh, this should be fixed. It is still left true from testing, but I wanted to set it to false again.  We still format docstrings with docformatter, since ruff only can format docstring-code lines and not the whole docsting.

KelvinChung2000 commented:
Something came to my mind: variable naming. Should we also enforce that?

EverythingElseWasAlreadyTaken commented:
> Something came to my mind: variable naming. Should we also enforce that?

I also thought about this, but it will be a mess to fix in the first place, since there is really no constituency currently.

IAmMarcelJung commented:
> Something came to my mind: variable naming. Should we also enforce that?

At some point we should do this, but yeah, probably not now.

EverythingElseWasAlreadyTaken commented:
I've added a lot more rules to the ruff.toml, which were kinda easy to apply and make sense to us. 
I've also added a list of rules we should apply in the future, but these mostly need a lot of manual changes, which just takes time. I'll open an issue for this, after this PR is merged.

EverythingElseWasAlreadyTaken commented:
Rebase on #384
